src/main/java/org/folio/rest/impl/InvoiceHelper.java

```diff
@@ -149,15 +176,355 @@
+      throw new HttpException(500, NO_INVOICE_LINES_ERROR_MSG);
+    }
+  }
+
+  /**
+   * Prepares a new voucher or updates existing one for further processing
+   *
+   * @param invoice {@link Invoice} to be approved on the basis of which the voucher is prepared
+   * @return completable future with {@link Voucher} on success
+   */
+  private CompletableFuture<Voucher> prepareVoucher(Invoice invoice) {
+    return findExistingVoucher(invoice.getId())
+      .thenCompose(vouchers -> {
+        if (isNotEmpty(vouchers)){
+          Voucher voucher = vouchers.get(0);
+          return updateExistingVoucherState(invoice, voucher);
         }
+        return buildNewVoucher(invoice);
       });
   }
 
+  private CompletableFuture<List<Voucher>> findExistingVoucher(String invoiceId) {
+    return voucherHelper.getVouchers(1, 0, String.format(QUERY_BY_INVOICE_ID, invoiceId))
+      .thenApply(VoucherCollection::getVouchers);
+  }
+
+  /**
+   * Updates state of existing {@link Voucher} linked with processed {@link Invoice}
+   *
+   * @param invoice invoice {@link Invoice} to be approved
+   * @param voucher {@link Voucher} from voucher-storage related to processed invoice
+   * @return completable future with {@link Voucher}
+   */
+  private CompletableFuture<Voucher> updateExistingVoucherState(Invoice invoice, Voucher voucher) {
+    voucher.setInvoiceCurrency(invoice.getCurrency());
+    setDefaultRequiredFields(voucher);
+
+    return getSystemCurrency()
+      .thenCompose(systemCurrency -> getExchangeRate(voucher.withSystemCurrency(systemCurrency)))
+      .thenApply(voucher::withExchangeRate);
+  }
+
+  //TODO Start using real information to create a voucher when it becomes known where to get these values from.
+  private void setDefaultRequiredFields(Voucher voucher) {
+    voucher.setAccountingCode(DEFAULT_ACCOUNTING_CODE);
+    voucher.setExportToAccounting(false);
+    voucher.setType(Voucher.Type.VOUCHER);
+    voucher.setStatus(Voucher.Status.AWAITING_PAYMENT);
+  }
+
+  /**
+   * Gets exchange rate based on {@link Voucher#invoiceCurrency} and {@link Voucher#systemCurrency} using default ExchangeRateProvider
+   *
+   * @param voucher {@link Voucher} for which the exchange rate factor is obtained
+   * @return {@link Double} value of exchange rate factor
+   */
+  private CompletableFuture<Double> getExchangeRate(Voucher voucher) {
+    return VertxCompletableFuture.supplyAsync(() -> MonetaryConversions.getExchangeRateProvider()
+      .getExchangeRate(voucher.getInvoiceCurrency(), voucher.getSystemCurrency())
+      .getFactor().doubleValue());
+  }
+
+  /**
+   * Build new {@link Voucher} based on processed {@link Invoice}
+   *
+   * @param invoice invoice {@link Invoice} to be approved
+   * @return completable future with {@link Voucher}
+   */
+  private CompletableFuture<Voucher> buildNewVoucher(Invoice invoice) {
+
+    Voucher voucher = new Voucher();
+    voucher.setInvoiceId(invoice.getId());
```

Comment:
For now this is fine - we'll eventually have to revisit this once we implement aggregate and batch vouchers which relate to multiple invoices.